src/components/TreeGrid/TreeGrid.tsx

```diff
@@ -0,0 +1,254 @@
+        let emptyArray = new Array();
+        emptyArray.push({
+            isSortable: false,
+            width: 18,
+            minWidth: 30,
+            fixedWidth: true
+        });
+        emptyArray.push({
+            dataType: DataTypeEnum.String,
+            valueMember: 'TreeId',
+            headerText: 'TreeId',
+            width: 0
+        });
+        columns = emptyArray.concat(columns);
+        return columns;
+    }
+
+    componentWillUpdate(nextProps, nextState) {
+        this.finalGridRows = getTreeRowsSelector(nextState, nextProps);
+    }
+
+    treeCellRenderer = ({ columnIndex, key, rowIndex, style }) => {
+        const rowData = this.finalGridRows[rowIndex]; 
+        const rowID: string = !rowData.TreeId ? '' : rowData.TreeId;
+        const indentSize = 20;
+        let indent = 0;
+        let level = getNodeLevel(rowData, this.finalGridRows);
+        if ((columnIndex === 0 || columnIndex === 2)) {
+
+            indent = level * indentSize;     
+        }
+        let shouldIndent: boolean = false;
+        if (columnIndex === 0) {
+            if (level === 0) {
+                shouldIndent = false;
+            } else if (style.left < indent) {
+                shouldIndent = true;
+            }
+        } else if (columnIndex === 2) {
+            if (level === 0) {
+                shouldIndent = false;
+            } else if (style.left < (indent + indentSize)) {
+                shouldIndent = true;
+            }
+        }
+        if (columnIndex === 1) {
+            return this.renderHiddenTreeCell(key, columnIndex, rowID);     
+        }
+        if (columnIndex === 2 && shouldIndent) {
+                style = {...style, width: style.width - indent};
+        }      
+        if (shouldIndent) {
+            style = {...style, left: style.left + indent};
+        }
+        if (columnIndex === 0) {
+            return this.renderActionTreeCell(key, rowIndex, rowData, style);
+        }
+        return this.renderBodyCell(columnIndex, key, rowIndex, rowData, style);     
+    }
+
+    renderHiddenTreeCell(key, columnIndex, rowData) {
+        const style = {
+            display: 'none'
+        };
+        return (
+        <div key={key} style={style}> 
+            {rowData}
+        </div>);
+    }
+
+    renderActionTreeCell(key, rowIndex: number, rowData: TreeNode, style) {
+        let actionsTooltip = rowData.IsExpanded ? 'Collapse' : 'Expand';
+        let iconName = rowData.IsExpanded ? 'icon-arrow_down' : 'icon-arrow_right';
+        let icon = null;
+        
+        if (rowData.leaves.length <= 0 && rowData.IsExpanded) {
+            icon = null;
+            actionsTooltip = null;
+        } else {
+            icon = <Icon iconName={iconName} className="expand-collapse-action-icon" />;
+        }
+        const rowClass = 'grid-row-' + rowIndex;
+        const onMouseEnter = () => { this.onMouseEnterCell(rowClass); };
```

Comment:
These arrow function may cause unnecessary rerenders. 
You could avoid them by using special component for the **div** you are returning and binding these functions.

Reply:
@hbagarich I made a new child component "CellElement". Can you just check it over real quickly to see if its properly implemented? 

Also when rendering the column element I am passing the result of ColumnElement function. Unsure if that is considered in the same context as passing the reference to the arrow function?

Reply:
Also I'm unsure if the @autobind is needed in that child component but I added it anyway. If it's redundant I can remove it